src/ufuncs.c

```diff
@@ -2261,30 +2311,49 @@ static void create_collection_func(char** args, npy_intp* dimensions, npy_intp*
         actual_type = GEOSGeomTypeId_r(ctx, g);
         if (actual_type == -1) {
           errstate = PGERR_GEOS_EXCEPTION;
+          destroy_geom_arr(ctx, geom_arr, i - 1);
+          destroy_geom_arr(ctx, temp_geoms, n_geoms - 1);
           goto finish;
         }
         if ((actual_type != expected_type) & (actual_type != alt_expected_type)) {
           errstate = PGERR_GEOMETRY_TYPE;
+          destroy_geom_arr(ctx, geom_arr, i - 1);
+          destroy_geom_arr(ctx, temp_geoms, n_geoms - 1);
           goto finish;
         }
       }
       g_copy = GEOSGeom_clone_r(ctx, g);
       if (g_copy == NULL) {
         errstate = PGERR_GEOS_EXCEPTION;
+        destroy_geom_arr(ctx, geom_arr, i - 1);
+        destroy_geom_arr(ctx, temp_geoms, n_geoms - 1);
         goto finish;
       }
-      geoms[n_geoms] = g_copy;
+      temp_geoms[n_geoms] = g_copy;
       n_geoms++;
     }
-    GEOSGeometry* ret_ptr = GEOSGeom_createCollection_r(ctx, type, geoms, n_geoms);
-    OUTPUT_Y;
+    geom_arr[i] = GEOSGeom_createCollection_r(ctx, type, temp_geoms, n_geoms);
+    if (geom_arr[i] == NULL) {
+      // We will have a memory leak now (https://trac.osgeo.org/geos/ticket/1111)
```

Comment:
```suggestion
      // We may have a memory leak now (https://trac.osgeo.org/geos/ticket/1111)
```

(We know polygons are problematic; not sure about collections)